ocaml-lsp-server/src/ocamlformat_rpc.ml

```diff
@@ -0,0 +1,172 @@
+open Import
+
+type process =
+  { pid : int
+  ; input : in_channel
+  ; output : out_channel
+  ; io_thread : Scheduler.thread
+  ; client : Ocamlformat_rpc_lib.client
+  }
+
+type state =
+  | Binary_not_found
+  | Stopped
+  | Running of process
+  | Errored
+
+type t = state ref
+
+let supported_versions = [ "v1" ]
+
+let pick_client ~pid input output io_thread =
+  let open Fiber.O in
+  match
+    Scheduler.async io_thread (fun () ->
+        Ocamlformat_rpc_lib.pick_client ~pid input output supported_versions)
+  with
+  | Error `Stopped -> Fiber.return @@ Error (`Msg "Thread stopped")
+  | Ok res -> (
+    let+ res = Scheduler.await_no_cancel res in
+    match res with
+    | Ok s -> s
+    | Error e -> Exn_with_backtrace.reraise e)
+
+let configure process =
+  let open Fiber.O in
+  let { io_thread; client; _ } = process in
+  (* We ask for 64 columns formatting as this appear to be the maximum size of
+     VScode popups. TODO We should probably allow some flexibility for other
+     editors that use the server. *)
+  match
+    Scheduler.async io_thread (fun () ->
+        match
+          Ocamlformat_rpc_lib.config
+            [ ("module-item-spacing", "compact"); ("margin", "64") ]
+            client
+        with
+        | Error (`Msg msg) ->
+          Log.log ~section:"ocamlformat" (fun () ->
+              Log.msg "An error occured while configuring ocamlformat"
+                [ ("msg", `String msg) ])
+        | Ok () -> ())
+  with
+  | Error `Stopped -> Fiber.return ()
+  | Ok res -> (
+    let+ res = Scheduler.await_no_cancel res in
+    match res with
+    | Ok s -> s
+    | Error e -> Exn_with_backtrace.reraise e)
+
+let start_exn ?io_thread t =
+  let prog = "ocamlformat-rpc" in
+  match Bin.which prog with
+  | None ->
+    t := Binary_not_found;
+    Fiber.return @@ Error `No_process
+  | Some bin_path -> (
+    let bin = Fpath.to_string bin_path in
+    let argv = [| bin |] in
+    let pid, stdout, stdin =
+      let stdin_i, stdin_o = Unix.pipe () in
+      let stdout_i, stdout_o = Unix.pipe () in
+      let pid = Unix.create_process bin argv stdin_i stdout_o Unix.stderr in
+      (pid, stdout_i, stdin_o)
+    in
+    let open Fiber.O in
+    let input = Unix.in_channel_of_descr stdout in
+    let output = Unix.out_channel_of_descr stdin in
+    let* io_thread =
+      (* Don't create a new thread if there is already one. *)
+      (* This can happen if the Ocamlformat server crashed. *)
```

Comment:
I'd prefer not to reuse the threads between processes unless there's a concrete performance improvement that is totally necessary. The issue is that we might be stuck arbitrarily long waiting for existing tasks to complete. If we create a new thread, we'll always get new responses promptly.